cmd/service-controller/service_sync.go

```diff
@@ -36,33 +44,27 @@ func (c *Controller) serviceSyncDefinitionsUpdated(definitions map[string]types.
 			if _, ok := c.byOrigin[service.Origin]; !ok {
 				c.byOrigin[service.Origin] = make(map[string]types.ServiceInterface)
 			}
-			c.byOrigin[service.Origin][name] = service
+			// track by origin if not locally exposed
+			if _, ok := c.localServices[name]; !ok {
```

Comment:
I don't think this check should be here. The `definitions` map holds the latest local service definitions (i.e. the services as defined in the configmap). The `localServices` map holds the previous set of service definitions.

E.g. in a scenario where the local service is unexposed and immediately after we get a service sync update from another site containing that service, the definition in the configmap would be updated to have the new origin. However the localServices will still hold the stale definition.

I believe the origin in the records in  `definitions` is authoritative here. (If we wanted we could ensure that the definition can never have both an originat that is non-null and not "annotations" *and* a target defined. The presence of a target would invalidate an explicit origin.

Reply:
Good catch. Will remove check and do additional testing including annotations, full mesh, etc.